design/helm.md

```diff
@@ -0,0 +1,145 @@
+1. Generate a bundle including the chart + containers
+
+In the airgapped case, install.sh will perform the following:
+1. Unpack the helm bundle on initialization
+1. Load docker containers as part of the initialization
+1. Unarchive the chart and install with the installer spec values during the install phase
+
+## Detailed Design
+
+The yaml design is informed by the [Flux Helm Operator CRD](https://github.com/fluxcd/helm-operator/blob/master/docs/references/helmrelease-custom-resource.md). 
+Another spec is [Rancher's Helm CRD](/var/lib/rancher/rke2/server/manifests/) that works with k3s and RKE2.
+
+The proposal is to implement option1 in the updated Kurl Installer Spec:
+```yaml
+apiVersion: "cluster.kurl.sh/v1beta1"
+kind: "Installer"
+metadata: 
+  name: "latest"
+spec: 
+  kubernetes: 
+    version: "latest"
+  weave: 
+    version: "latest"
+  containerd: 
+    version: "latest"
+  helm:                 # Optional Future Capability
+    version: "latest"
+  helmReleases:
+    # Basic Idea, Option 1
+    - release: nginx-ingress
+      namespace: kube-system    # Optional
+      chart: 
+        name: bitnami/nginx
+        repo: https://charts.bitnami.com/bitnami 
+        version: 8.5.1          # Optional - uses latest if not provided
+      values: |                 # Optional
+        ingress:
+          enabled: true
+          hostname: dan-dot-com
+        metrics: 
+          enabled: true
+    # Option 2 (Using Mapstructure?)
+    - release: postgres-web
+      chart: 
+        name: bitnami/postgresql
+        repo: https://charts.bitnami.com/bitnami 
+      values:
+        postgresqlUsername: dan
+        postgresqlPassword: opensesame
+    # Option 3 (Config Url)
+    - release: postgres-auth
+      chart: 
+        name: bitnami/postgresql
+        repo: https://charts.bitnami.com/bitnami 
+      values: https://gist.githubusercontent.com/DanStough/b4376b7a6aa5b73c6a647bff504212df/raw/829680f4a474720dc36b7a963d2d691305b8cdfb/postgres-auth.yaml
+  ...
+```
+
+### Assumptions
+1. Authenticated Registries are NOT supported by Kurl
+1. The join and upgrade functionality of Kurl is transparent to Helm installs. 
+
+### Design Decisions
+1. Re-install a chart: chart installs are not idempotent, so this will be treated as an upgrade.
+1. State between executions - the only place we want to keep a 
+1. Waiting - the (initial) default behavior of kurl will be to wait for resources to create. This will ensure that the cluster is in a known state before the user continues with the next set of installation (kotadm)
+1. RKE2 support: has built in helm chart support by putting the Rancher Helm CR in the directory `/var/lib/rancher/rke2/server/manifests/`, but we will initially manage helm support through the CLI to have one code path for all distros.
+ 
+### Limitations
+1. If someone changes the name of a helm 
+1. If a helm release cannot be upgraded, it will need to be fixed manually (no option to use --force)
```

Comment:
Helmfile has an interesting comment about being able to periodically sync and enforce Helm charts match declarative configuration. I wonder if we should consider such a thing.

Reply:
I think you'd need to move the periodic sync to a chron job in k8s (I didn't see anything about setting a timer or interval for helmfile) which could be done. I have this nagging feeling we are creating two sources of truth though, the installer and declarative helmfile (that I'm assuming we export) and that the two could get out of sync if people are trying to manage them locally.

Reply:
The intent here would definitely be to **not** have folks manipulate the helmfile after installation.